File: Dsa-day155.java
class Solution {
    public List<Integer> findAllPeople(int n, int[][] meetings, int firstPerson) {
        Arrays.sort(meetings,(a,b)-> a[2]-b[2]);

        int[] parent =new int[n];
        boolean[] secret = new boolean[n];
        for(int i=0;i<n; i++)parent[i]=i;

        secret[0] = secret[firstPerson]=true;
        for(int i=0;i<meetings.length;){
            int time =meetings[i][2];
            Set<Integer> group = new HashSet<>();

            int j=i;
            while(j<meetings.length && meetings[j][2] == time){
                union(meetings[j][0],meetings[j][1],parent);
                group.add(meetings[j][0]);
                group.add(meetings[j][1]);
                j++;
            }
            for(int p:group)
            if(secret[p])
            secret[find(p,parent)] = true;

            for(int p:group)
            secret[p]=secret[find(p,parent)];

            for(int p:group)
            parent[p]=p;

            i=j;
        }

        List<Integer> ans =new ArrayList<>();
        for(int i=0;i<n;i++){
            if(secret[i])ans.add(i);
        }
        return ans;
    }
        private int find (int x, int[] p){
            return p[x] == x?x : (p[x] =find(p[x],p));
        }
        private void union(int a, int b,int[] p){
            p[find(b,p)]=find(a,p);
        }
    
}

File: Dsa-day155.cpp
class Solution {
public:
    vector<int> findAllPeople(int n, vector<vector<int>>& meetings, int firstPerson) {
        map<int, vector<pair<int, int>>> meetingsByTime;
        for (auto& m : meetings) {
            meetingsByTime[m[2]].push_back({m[0], m[1]});
        }
        
        vector<int> parent(n);
        iota(parent.begin(), parent.end(), 0);
        
        parent[firstPerson] = 0;
        
        for (auto& [time, meetingsAtTime] : meetingsByTime) {
            for (auto& [x, y] : meetingsAtTime) {
                unionSets(parent, x, y);
            }
            
            for (auto& [x, y] : meetingsAtTime) {
                if (find(parent, x) != find(parent, 0)) {
                    parent[x] = x;
                }
                if (find(parent, y) != find(parent, 0)) {
                    parent[y] = y;
                }
            }
        }
        
        vector<int> result;
        for (int i = 0; i < n; i++) {
            if (find(parent, i) == find(parent, 0)) {
                result.push_back(i);
            }
        }
        
        return result;
    }
    
private:
    int find(vector<int>& parent, int x) {
        if (parent[x] != x) {
            parent[x] = find(parent, parent[x]);
        }
        return parent[x];
    }
    
    void unionSets(vector<int>& parent, int x, int y) {
        int px = find(parent, x);
        int py = find(parent, y);
        if (px != py) {
            parent[py] = px;
        }
    }
};